### PR TITLE
[SYCL] Use scheduler in queue shortcuts to avoid waiting for deps

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
Some queue shortcuts attempt to bypass the scheduler to avoid the
overhead of creating a command node. In cases where dependencies
couldn't be passed to the backend directly, there was a blocking wait
during submission.

This patch changes that behavior to bypass the scheduler only if all
dependency events can be used directly: host task dependencies have been
completed, others have been enqueued and there are no cross-context
dependencies.

### CODE REVIEW

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 1]
```diff
@@ -517,6 +518,8 @@ void Scheduler::cleanupDeferredMemObjects(BlockingT Blocking) {
 
   std::vector<std::shared_ptr<SYCLMemObjI>> ObjsReadyToRelease;
   {
+    static size_t count = 0;
+
```
Comment (KseniyaTikhomirova) [inline sycl/source/detail/scheduler/scheduler.cpp:522]: remove!

**Comment by KseniyaTikhomirova** [comment]
win failures are reproduced locally, will investigate

**Comment by KseniyaTikhomirova** [comment]
Win failure is observed in another pull request too. Reported here https://github.com/intel/llvm/issues/12011

**Comment by KseniyaTikhomirova** [comment]
@bso-intel , @sergey-semenov kindly ping

**Review by bso-intel — APPROVED** [review]

File: sycl/source/detail/queue_impl.cpp [hunk 2]
```diff
@@ -85,43 +141,48 @@ event queue_impl::memset(const std::shared_ptr<detail::queue_impl> &Self,
...
+    std::unique_lock<std::mutex> guard(MMutex, std::defer_lock);
+    EventImplPtr ExtraEventToWait = nullptr;
     if (isInOrder()) {
-      MLastEvent = ResEvent;
-      // We don't create a command group for usm commands, so set it to None.
-      // This variable is used to perform explicit dependency management when
-      // required.
-      MLastCGType = CG::CGTYPE::None;
+      guard.lock();
+      ExtraEventToWait = MGraph.lock() ? MGraphLastEventPtr : MLastEventPtr;
```
Comment (sergey-semenov) [inline sycl/source/detail/queue_impl.cpp:151]:
```suggestion
      ExtraEventToWait = MGraph.expired() ? MLastEventPtr : MGraphLastEventPtr;
```
This is probably better since you don't need the shared pointer here.

File: sycl/source/detail/queue_impl.cpp [hunk 3]
```diff
@@ -85,43 +141,48 @@ event queue_impl::memset(const std::shared_ptr<detail::queue_impl> &Self,
...
-
-    auto EventImpl = detail::getSyclObjImpl(ResEvent);
-    MemoryManager::fill_usm(Ptr, Self, Count, Value,
-                            getOrWaitEvents(DepEvents, MContext),
-                            &EventImpl->getHandleRef(), EventImpl);
-
-    if (MContext->is_host())
-      return MDiscardEvents ? createDiscardedEvent() : event();
-
+    std::unique_lock<std::mutex> guard(MMutex, std::defer_lock);
```
Comment (sergey-semenov) [inline sycl/source/detail/queue_impl.cpp:147]: Not a blocking comment for this PR since this was here before, but there seems to be a lot of code duplication between these queue shortcut functions. It would be nice to move out the common part to a separate function.

**Review by sergey-semenov — COMMENTED** [review]
LGTM overall, mostly just minor comments and nitpicks.

**Comment by againull** [comment]
I'm sorry to point that out, but I accidentally found out in this PR's CI https://github.com/intel/llvm/pull/12069  that SchedulerTest.InOrderQueueCrossDepsShortcutFuncs hangs flakily with low rate.
Can be reproduced locally like this (may need to be executed several times to see the hang):
for i in {1..1000}; do echo Test $i;  GTEST_SHUFFLE=0 GTEST_TOTAL_SHARDS=20 GTEST_SHARD_INDEX=18  ./tools/sycl/unittests/scheduler/SchedulerTests; done

**Comment by KseniyaTikhomirova** [comment]
> I'm sorry to point that out, but I accidentally found out in this PR's CI #12069 that SchedulerTest.InOrderQueueCrossDepsShortcutFuncs hangs flakily with low rate. Can be reproduced locally like this (may need to be executed several times to see the hang): for i in {1..1000}; do echo Test $i; GTEST_SHUFFLE=0 GTEST_TOTAL_SHARDS=20 GTEST_SHARD_INDEX=18 ./tools/sycl/unittests/scheduler/SchedulerTests; done

don't be sorry! good to find problems before merge. Will take a look, thanks.

File: sycl/source/detail/queue_impl.cpp (hunk 2 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.cpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/source/detail/queue_impl.cpp [hunk 4]
```diff
@@ -56,6 +111,7 @@ static event prepareSYCLEventAssociatedWithQueue(
 }
 
 static event createDiscardedEvent() {
+
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.cpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/source/detail/queue_impl.hpp [hunk 5]
```diff
@@ -714,29 +719,23 @@ class queue_impl {
...
-
-      if (NeedSeparateDependencyMgmt)
-        Handler.depends_on(MLastEvent);
+      std::lock_guard<std::mutex> Lock{MMutex};
+      // This dependency is needed for the following purposes:
+      //    - host tasks is handled by runtime and could not be implicitly
+      //    synchronized by backend.
+      //    - to prevent 2nd kernel enqueue when 1st kernel is blocked by host
+      //    task. This dependency allows to build enqueue order in RT but will
+      //    be not passed to backend. Look at getPIEvents in Command.
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.hpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/unittests/scheduler/InOrderQueueHostTaskDeps.cpp [hunk 6]
```diff
@@ -32,23 +33,137 @@ inline pi_result redefinedEventsWait(pi_uint32 num_events,
...
+  ExecutedCommands.clear();
+  sycl::unittest::PiMock Mock;
+  Mock.redefineBefore<detail::PiApiKind::piEnqueueKernelLaunch>(
+      customEnqueueKernelLaunch);
+  Mock.redefineBefore<detail::PiApiKind::piextUSMEnqueueMemset>(
+      customextUSMEnqueueMemset);
+
+  sycl::platform Plt = Mock.getPlatform();
+  if (Plt.is_host()) {
+    std::cout << "Not run due to host-only environment\n";
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/unittests/scheduler/InOrderQueueHostTaskDeps.cpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/unittests/scheduler/InOrderQueueHostTaskDeps.cpp [hunk 7]
```diff
@@ -32,23 +33,137 @@ inline pi_result redefinedEventsWait(pi_uint32 num_events,
...
+  auto buf = sycl::malloc_shared<int>(1, InOrderQueue);
+  event Evt = InOrderQueue.submit(
+      [&](sycl::handler &CGH) { CGH.memset(buf, 0, sizeof(buf[0])); });
+  InOrderQueue.submit([&](sycl::handler &CGH) { CGH.host_task([=] {}); })
+      .wait();
+
+  EXPECT_EQ(GEventsWaitCounter, 1u);
+}
+
+enum CommandType { KERNEL = 1, MEMSET = 2 };
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/unittests/scheduler/InOrderQueueHostTaskDeps.cpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/source/detail/queue_impl.cpp [hunk 8]
```diff
@@ -26,9 +26,64 @@
 namespace sycl {
 inline namespace _V1 {
 namespace detail {
-
 std::atomic<unsigned long long> queue_impl::MNextAvailableQueueID = 0;
 
+std::vector<sycl::detail::pi::PiEvent>
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.cpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/source/detail/queue_impl.cpp [hunk 9]
```diff
@@ -26,9 +26,64 @@
...
+    AddEvent(ExtraDepEvent);
+  for_each(DepEvents.begin(), DepEvents.end(),
+           [&RetPiEvents, &AddEvent](const sycl::event &Event) {
+             auto EventImpl = detail::getSyclObjImpl(Event);
+             return AddEvent(EventImpl);
+           });
+  return RetPiEvents;
+}
+
+bool isEventsReady(const std::vector<sycl::event> &DepEvents,
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.cpp]: https://github.com/intel/llvm/pull/11758/commits/44eabe55c1b6557652c58c8d16cb44aaa6dc2bac

File: sycl/source/detail/scheduler/commands.cpp [hunk 10]
```diff
@@ -513,6 +513,7 @@ Command::Command(
   MEvent->setCommand(this);
   MEvent->setContextImpl(MQueue->getContextImplPtr());
   MEvent->setStateIncomplete();
+  MEvent->setProducesPiEvent(producesPiEvent());
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp]: done in https://github.com/intel/llvm/pull/11758/commits/93bee37c6747a6ae6295107f18204b7c85b81482

File: sycl/source/detail/queue_impl.cpp (hunk 3 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.cpp:150]: to be considered and implemented separately :-)

**Comment by sergey-semenov** [comment]
@bso-intel @intel/llvm-reviewers-runtime Could you please have a look at the latest few commits? 

The workaround in `Command::producesPiEvent` is still needed so I restored it, but there's no need to propagate that value to the event when we can just check the `pi_event` handle there to achieve the same result.

The barrier changes were also reverted since the extension spec seems to be pretty clear that "If waitList is empty, then the barrier has no effect".

File: sycl/source/detail/queue_impl.cpp [hunk 11]
```diff
@@ -26,9 +26,59 @@
...
+      return;
+    RetPiEvents.push_back(EventImpl->getHandleRef());
+  };
+  if (ExtraDepEvent)
+    AddEvent(ExtraDepEvent);
+  for_each(DepEvents.begin(), DepEvents.end(),
+           [&RetPiEvents, &AddEvent](const sycl::event &Event) {
+             auto EventImpl = detail::getSyclObjImpl(Event);
+             return AddEvent(EventImpl);
+           });
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:46]:
```suggestion
  for(sycl::event &Event : DepEvents)
    AddEvent(detail::getSyclObjImpl(Event);
```

File: sycl/source/detail/queue_impl.cpp [hunk 12]
```diff
@@ -26,9 +26,59 @@
...
+  return RetPiEvents;
+}
+
+static bool isEventsReady(const std::vector<sycl::event> &DepEvents,
+                          const EventImplPtr &ExtraDepEventPtr,
+                          ContextImplPtr Context) {
+  auto CheckEvent = [&Context](const EventImplPtr &SyclEventImplPtr) {
+    // throwaway events created with empty constructor will not have a
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:54]:
```suggestion
    // Throwaway events created with empty constructor will not have a
```

File: sycl/source/detail/queue_impl.cpp [hunk 13]
```diff
@@ -26,9 +26,59 @@
...
+    AddEvent(ExtraDepEvent);
+  for_each(DepEvents.begin(), DepEvents.end(),
+           [&RetPiEvents, &AddEvent](const sycl::event &Event) {
+             auto EventImpl = detail::getSyclObjImpl(Event);
+             return AddEvent(EventImpl);
+           });
+  return RetPiEvents;
+}
+
+static bool isEventsReady(const std::vector<sycl::event> &DepEvents,
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:50]:
```suggestion
static bool areEventsReady(const std::vector<sycl::event> &DepEvents,
```

File: sycl/source/detail/queue_impl.cpp [hunk 14]
```diff
@@ -26,9 +26,59 @@
...
+    // context (which is set lazily) calling getContextImpl() would set that
+    // context, which we wish to avoid as it is expensive.
+    if (!SyclEventImplPtr->isContextInitialized() &&
+        !SyclEventImplPtr->is_host()) {
+      return true;
+    }
+    if (SyclEventImplPtr->is_host()) {
+      return SyclEventImplPtr->isCompleted();
+    }
+    // Cross-context dependencies can't be passed directly.
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:64]: They way the comment is worded right now I'd expect a deadlock here... Can we be more elaborate here? My best guess would be that this entire routine is an "optimization" path and we don't need to provide a precise "false" answer, only a precise "true" one. If so, that should into the documentation of the function.

File: sycl/source/detail/queue_impl.cpp [hunk 15]
```diff
@@ -85,49 +135,55 @@ event queue_impl::memset(const std::shared_ptr<detail::queue_impl> &Self,
...
+    std::unique_lock<std::mutex> guard(MMutex, std::defer_lock);
+    EventImplPtr ExtraEventToWait = nullptr;
     if (isInOrder()) {
-      MLastEvent = ResEvent;
-      // We don't create a command group for usm commands, so set it to None.
-      // This variable is used to perform explicit dependency management when
-      // required.
-      MLastCGType = CG::CGTYPE::None;
+      guard.lock();
+      ExtraEventToWait = MGraph.expired() ? MLastEventPtr : MGraphLastEventPtr;
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:152]: Can we just `push_back` it into the `DepEvents`?

File: sycl/source/detail/queue_impl.cpp [hunk 16]
```diff
@@ -85,49 +135,55 @@ event queue_impl::memset(const std::shared_ptr<detail::queue_impl> &Self,
...
+        return createDiscardedEvent();
+      }
+
+      event ResEvent = prepareSYCLEventAssociatedWithQueue(Self);
+      auto EventImpl = detail::getSyclObjImpl(ResEvent);
+      MemoryManager::fill_usm(Ptr, Self, Count, Value,
+                              getPIEvents(DepEvents, ExtraEventToWait),
+                              &EventImpl->getHandleRef(), EventImpl);
+      if (MContext->is_host())
+        return MDiscardEvents ? createDiscardedEvent() : event();
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:168]: Why can we have `MDiscardEvents && !MHasDiscardEVentsSupport` at the same time?

File: sycl/source/detail/queue_impl.cpp [hunk 17]
```diff
@@ -184,191 +240,207 @@ event queue_impl::memcpy(const std::shared_ptr<detail::queue_impl> &Self,
...
   EventImplPtr EImpl = getSyclObjImpl(Event);
   assert(EImpl && "Event implementation is missing");
   auto *Cmd = static_cast<Command *>(EImpl->getCommand());
   if (!Cmd) {
-    // if there is no command on the event, we cannot track it with MEventsWeak
-    // as that will leave it with no owner. Track in MEventsShared only if we're
-    // unable to call piQueueFinish during wait.
+    // if there is no command on the event, we cannot track it with
+    // MEventsWeak as that will leave it with no owner. Track in MEventsShared
+    // only if we're unable to call piQueueFinish during wait.
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:443]: Unrelated changes here and below?

**Review by aelovikov-intel — COMMENTED** [review]
Some stylistic comments.

File: sycl/source/detail/queue_impl.cpp (hunk 16 above)
Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp:168]: `MDiscardEvents` indicates whether the property was passed to the queue (should we discard SYCL events or not), 'MHasDiscardEventsSupport' is a necessary condition for omitting PI/UR events (this requires MDiscardEvents to be true and the queue being in order).

I agree that the naming is incredibly confusing here and even the comment for 'MHasDiscardEventsSupport` doesn't help with that. I'll follow up with another PR to address that.

File: sycl/source/detail/queue_impl.cpp (hunk 17 above)
Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp]: Reverted

File: sycl/source/detail/queue_impl.cpp (hunk 15 above)
Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp:148]: This is to avoid copying the vector since it's passed by const reference.

File: sycl/source/detail/queue_impl.cpp (hunk 14 above)
Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp]: I think the actual problem here was the misleading function name. Changed it to `canBypassScheduler` and made small updates to the comments.

**Comment by sergey-semenov** [comment]
The remaining test failure is actually due to an existing runtime bug (the test currently matches this incorrect behavior). Opened https://github.com/intel/llvm/pull/12386

File: sycl/source/detail/queue_impl.cpp (hunk 15 above)
Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp]: Since we have another source of extra events in the form of the `sycl_ext_oneapi_in_order_queue_events` extension now, I ended up reusing the same copied vector that's used for that.

**Review by kbenzie — APPROVED** [review]
UR LGTM

**Review by bso-intel — APPROVED** [review]

File: sycl/source/detail/queue_impl.cpp [hunk 18]
```diff
@@ -189,60 +239,57 @@ event queue_impl::memcpy(const std::shared_ptr<detail::queue_impl> &Self,
...
+        return MDiscardEvents ? createDiscardedEvent() : event();
+      if (isInOrder()) {
+        auto &EventToStoreIn =
+            MGraph.lock() ? MGraphLastEventPtr : MLastEventPtr;
+        EventToStoreIn = EventImpl;
+      }
+      // Track only if we won't be able to handle it with piQueueFinish.
+      if (MEmulateOOO)
+        addSharedEvent(ResEvent);
+      return discard_or_return(ResEvent);
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:288]: Can we outline it into a helper? I'd prefer the refactoring to land before this PR, unless the fix is really urgent.

Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp]: Done

Comment (sergey-semenov) [reply sycl/source/detail/queue_impl.cpp]: Opened https://github.com/intel/llvm/pull/12474

**Comment by sergey-semenov** [comment]
@aelovikov-intel Could you please have another look?

File: sycl/source/detail/queue_impl.cpp [hunk 19]
```diff
@@ -273,55 +297,90 @@ void queue_impl::addSharedEvent(const event &Event) {
   MEventsShared.push_back(Event);
 }
 
-template <typename MemOpFuncT, typename... MemOpArgTs>
+static bool
+areEventsSafeForSchedulerBypass(const std::vector<sycl::event> &DepEvents,
+                                ContextImplPtr Context) {
+  auto CheckEvent = [&Context](const sycl::event &Event) {
+    const EventImplPtr &SyclEventImplPtr = detail::getSyclObjImpl(Event);
+    // Events that don't have an initialized context are throwaway evemts that
```
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:305]:
```suggestion
    // Events that don't have an initialized context are throwaway events that
```

**Review by aelovikov-intel — APPROVED** [review]
Stylistically it's good, the comments are very helpful as well.

I've reviewed the logic to the best of my abilities/knowledge in the area but the latter is very limited and I might have missed something. That said, you're the expert in the area and I don't know if anybody available could help with the review, so let's merge it.

**Comment by sergey-semenov** [comment]
Last commit was a comment typo fix, merging based on the previous test run.

**Comment by martygrant** [comment]
@KseniyaTikhomirova @sergey-semenov there is a [post-commit](https://github.com/intel/llvm/actions/runs/7697466128) failure for this PR which should be addressed, as the error has occurred in a [more recent PR](https://github.com/intel/llvm/pull/12489). The errors are:

`/Users/runner/work/llvm/llvm/src/sycl/source/detail/queue_impl.hpp:747:64: error: unused parameter 'Type' [-Werror,-Wunused-parameter] void finalizeHandler(HandlerType &Handler, const CG::CGTYPE &Type,`

`/__w/llvm/llvm/src/sycl/source/detail/queue_impl.cpp:325:24: error: lambda capture 'Context' is not used [-Werror,-Wunused-lambda-capture] 325 |                      [&Context, &CheckEvent](const sycl::event &Event) {`

**Comment by sergey-semenov** [comment]
@martygrant Thanks for letting me know, opened https://github.com/intel/llvm/pull/12523

**Comment by aelovikov-intel** [comment]
It seems there are remaining warnings caused by this (https://github.com/intel/llvm/actions/runs/7701911776/job/20989034380):

```
/__w/llvm/llvm/src/sycl/unittests/scheduler/InOrderQueueHostTaskDeps.cpp:94:7: error: 'unique_lock' may not intend to support class template argument deduction [-Werror,-Wctad-maybe-unsupported]
   94 |       std::unique_lock lk(CvMutex);
      |       ^
```
